can/scripts/viewer.py

```diff
@@ -0,0 +1,602 @@
+                # node ID and message length
+                if canopen_function_code == 0x080:
+                    # Check if the length is valid
+                    if msg.dlc in canopen_function_codes[canopen_function_code]:
+                        # Make sure the length and node ID combination is valid
+                        if (msg.dlc == 0 and canopen_node_id == 0) or (msg.dlc == 8 and 1 <= canopen_node_id <= 127):
+                            canopen_function_code_string = canopen_function_codes[canopen_function_code][msg.dlc]
+                elif (canopen_function_code == 0x000 or canopen_function_code == 0x100) and \
+                        (canopen_node_id != 0 or msg.dlc not in canopen_function_codes[canopen_function_code]):
+                    # It is not a CANopen message, as the node ID is not added to these command
+                    canopen_function_code_string = None
+                else:
+                    if isinstance(canopen_function_codes[canopen_function_code], dict):
+                        # Make sure the message has the defined length
+                        if msg.dlc in canopen_function_codes[canopen_function_code]:
+                            canopen_function_code_string = canopen_function_codes[canopen_function_code][msg.dlc]
+                    # These IDs do not have a fixed length
+                    else:
+                        # Make sure the node ID is valid
+                        if 1 <= canopen_node_id <= 127:
+                            canopen_function_code_string = canopen_function_codes[canopen_function_code]
+
+                # Now determine set the node ID string
+                if canopen_function_code_string:
+                    if 1 <= canopen_node_id <= 127:  # Make sure the node ID is valid
+                        canopen_node_id_string = '0x{0:02X}'.format(canopen_node_id)
+                    elif canopen_function_code == 0x000:
+                        # The NMT command sends the node ID as the second byte, except when it is 0,
+                        # then the command is sent to all nodes
+                        if msg.data[1] == 0:
+                            canopen_node_id_string = 'ALL'
+                        elif 1 <= msg.data[1] <= 127:
+                            canopen_node_id_string = '0x{0:02X}'.format(msg.data[1])
+                        else:
+                            # It not a valid NMT command, as the node ID is not valid
+                            canopen_function_code_string = None
+            elif (msg.arbitration_id == 0x7E4 or msg.arbitration_id == 0x7E5) and \
+                    msg.dlc in canopen_function_codes[msg.arbitration_id]:
+                # Check if it is the LSS commands
+                canopen_function_code_string = canopen_function_codes[msg.arbitration_id][msg.dlc]
+
+        return canopen_function_code_string, canopen_node_id_string
+
+    def draw_can_bus_message(self, msg, sorting=False):
```

Comment:
CAN FD should also be supported eventually, but we could add that later. It's going to be tricky since the data string would have to span multiple lines.

Reply:
Well, we could display the data [ellipsized](https://en.wikipedia.org/wiki/Ellipsis#Computer_representations) for now.

Reply:
Error frames are now colored red. I have not added support for CAN FD, as I have no way to test it atm.